File: projects/ladderbet/ladder/model_accuracy.py
"""Read-only model history. Frozen pre-event predictions, never actual wagers.

One preferred side per event/market is scored separately from the complete
priced-side audit. Forecasts without a sportsbook quote are tracked as well.
Missing scores/statistics stay pending; historical forecasts are never invented.
"""
from __future__ import annotations

import json
import math
from collections import defaultdict
from datetime import datetime, timezone
from pathlib import Path

VERSION = "2026-09-06-ev2"


def instant(value):
    try:
        if "T" not in str(value):
            return None
        d = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        return d.astimezone(timezone.utc) if d.tzinfo else None
    except (TypeError, ValueError):
        return None


def number(value):
    try:
        n = float(value)
        return n if math.isfinite(n) else None
    except (TypeError, ValueError):
        return None


def load(path):
    path = Path(path)
    # Fail a refresh on corrupt history instead of silently erasing it.
    return json.loads(path.read_text()) if path.exists() else {"records": {}}


def save(path, value):
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(value, indent=1, allow_nan=False) + "\n")
    tmp.replace(path)


def key(row):
    base = ":".join(str(row.get(k) or "") for k in
                    ("kind", "league", "event_id", "player", "market", "side"))
    return base + (":" + str(row.get("line")) if row.get("kind") == "call" and row.get("player") else "")


def group(row):
    return ":".join(str(row.get(k) or "") for k in
                    ("league", "event_id", "player", "market"))


def record(log, rows, now=None):
    now = instant(now) if now is not None else datetime.now(timezone.utc)
    records = log.setdefault("records", {})
    chosen = {group(r) for r in records.values() if r.get("selected")}
    # The first selected side is immutable, including its line, tier and price.
    for source in sorted(rows, key=lambda r: -(number(r.get("edge")) or 0)):
        r = dict(source)
        k = key(r)
        if k in records:
            continue
        start = instant(r.get("start"))
        captured = instant(r.get("captured_at")) if r.get("captured_at") else now
        if not now or not start or not captured or captured >= start or captured > now:
            continue
        if r.get("completed") and not r.get("captured_at"):
            continue
        if not r.get("event_id") or r.get("kind") not in {"call", "game", "prop"}:
            continue
        if r["kind"] == "call":
            probability = number(r.get("probability"))
            if number(r.get("price")) in (None, 0) or probability is None or not 0 <= probability <= 1:
                continue
            r["probability"] = probability
            r["selected"] = group(r) not in chosen
            chosen.add(group(r))
        r.update(id=k, captured_at=captured.isoformat(), result="Pending",
                 version=r.get("version") or VERSION)
        previous = r.pop("prior_result", None)
        if previous in {"Win", "Loss", "Push", "Void"} and r["kind"] == "call":
            price = float(r["price"])
            units = (price/100 if price > 0 else 100/-price) if previous == "Win" else -1 if previous == "Loss" else 0
            r.update(result=previous, units=round(units, 6))
        r.pop("completed", None)
        records[k] = r
    return log


def settle(log, results, now=None):
    stamp = now or datetime.now(timezone.utc).isoformat()
    for r in log.get("records", {}).values():
        if r.get("result") != "Pending":
            continue
        g = results.get(str(r["event_id"])) or {}
        if g.get("canceled"):
            r.update(result="Void", graded_at=stamp, units=0)
            continue
        if not g.get("completed"):
            continue
        h, a = number(g.get("home_score")), number(g.get("away_score"))
        if r["kind"] == "game":
            if h is None or a is None:
                continue
            r.update(result="Graded", actual_margin=h-a, actual_total=h+a,
                     final_home=h, final_away=a, graded_at=stamp)
            continue
        actual = None
        if r.get("player"):
            actual = number((g.get("stats") or {}).get(r.get("stat_key")))
            if actual is None:
                # A missing player/stat is not evidence of a zero or a loss.
                continue
            if r["kind"] == "prop":
                r.update(result="Graded", actual=actual, graded_at=stamp)
                continue
            line = number(r.get("line"))
            if r.get("side") in {"yes", "no"}:
                delta = 1 if actual > 0 else -1
                if r["side"] == "no":
                    delta = -delta
            elif line is not None and r.get("side") in {"over", "under"}:
                delta = (actual-line) * (1 if r["side"] == "over" else -1)
            else:
                continue
        elif h is not None and a is not None:
            if r["market"] == "ML":
                if r.get("three_way"):
                    winner = "home" if h > a else "away" if a > h else "draw"
                    delta = 1 if r["side"] == winner else -1
                else:
                    delta = (h-a) * (1 if r["side"] == "home" else -1)
            elif r["market"] == "ATS" and number(r.get("line")) is not None:
                # Normalized line is always the selected team's handicap.
                delta = (h-a if r["side"] == "home" else a-h) + float(r["line"])
            elif r["market"] == "TOTAL" and number(r.get("line")) is not None:
                delta = (h+a-float(r["line"])) * (1 if r["side"] == "over" else -1)
            else:
                continue
        else:
            continue
        outcome = "Win" if delta > 1e-9 else "Loss" if delta < -1e-9 else "Push"
        price = float(r["price"])
        units = (price/100 if price > 0 else 100/-price) if outcome == "Win" else -1 if outcome == "Loss" else 0
        r.update(result=outcome, units=round(units, 6), actual=actual,
                 final_home=h, final_away=a, graded_at=stamp)
    return log


def mean(values):
    v = [n for x in values if (n := number(x)) is not None]
    return round(sum(v)/len(v), 5) if v else None


def call_stats(rows):
    wins = sum(r["result"] == "Win" for r in rows)
    losses = sum(r["result"] == "Loss" for r in rows)
    pushes = sum(r["result"] == "Push" for r in rows)
    voids = sum(r["result"] == "Void" for r in rows)
    settled = wins+losses+pushes
    decided = [r for r in rows if r["result"] in {"Win", "Loss"}]
    units = sum(r.get("units") or 0 for r in rows if r["result"] != "Pending")
    return {"logged": len(rows), "wins": wins, "losses": losses, "pushes": pushes,
            "voids": voids, "settled": settled, "pending": len(rows)-settled-voids,
            "win_rate": wins/(wins+losses) if wins+losses else None,
            "units": round(units, 3), "roi": units/settled if settled else None,
            "brier": mean([(r["probability"]-(r["result"] == "Win"))**2 for r in decided
                           if number(r.get("probability")) is not None])}


def buckets(rows, field):
    groups = defaultdict(list)
    for r in rows:
        groups[str(r.get(field) or "Unknown")].append(r)
    return {k: call_stats(v) for k, v in sorted(groups.items())}


def report(log, label="Model accuracy"):
    rows = list(log.get("records", {}).values())
    calls = [r for r in rows if r["kind"] == "call"]
    picks = [r for r in calls if r.get("selected")]
    games = [r for r in rows if r["kind"] == "game"]
    finals = [r for r in games if r["result"] == "Graded"]
    winner = [r for r in finals if number(r.get("margin")) not in (None, 0) and r["actual_margin"] != 0]
    ats = [r for r in finals if number(r.get("margin")) is not None and number(r.get("market_spread")) is not None
           and abs(r["margin"]+r["market_spread"]) > .25 and abs(r["actual_margin"]+r["market_spread"]) > 1e-9]
    totals = [r for r in finals if number(r.get("total")) is not None and number(r.get("market_total")) is not None
              and abs(r["total"]-r["market_total"]) > .25 and abs(r["actual_total"]-r["market_total"]) > 1e-9]
    matched_margin = [r for r in finals if number(r.get("margin")) is not None and number(r.get("market_spread")) is not None]
    matched_total = [r for r in finals if number(r.get("total")) is not None and number(r.get("market_total")) is not None]
    props = [r for r in rows if r["kind"] == "prop"]
    prop_groups = defaultdict(list)
    for r in props:
        prop_groups[r["market"]].append(r)
    calibration = []
    for i in range(10):
        rs = [r for r in picks if r["result"] in {"Win", "Loss"} and
              number(r.get("probability")) is not None and min(9, int(r["probability"]*10)) == i]
        if rs:
            calibration.append({"bucket": f"{i*10}–{(i+1)*10}%", "n": len(rs),
                                "predicted": mean([r["probability"] for r in rs]),
                                "actual": mean([r["result"] == "Win" for r in rs])})
    def rate(rs, fn):
        return {"n": len(rs), "correct": sum(bool(fn(r)) for r in rs),
                "accuracy": mean([bool(fn(r)) for r in rs])}
    return {"schema": 1, "generated_at": datetime.now(timezone.utc).isoformat(), "label": label,
            "method": "First pre-event snapshot, frozen line and price. All model picks, including PASS/AVOID and unstaked picks. One preferred side per event/market in pick statistics; the complete priced-side audit is separate. One hypothetical unit per pick; no actual wagers are created. Missing results remain pending.",
            "overall": call_stats(picks), "all_calls": call_stats(calls),
            "by_tier": buckets(picks, "tier"), "by_market": buckets(picks, "market"),
            "by_version": buckets(picks, "version"), "calibration": calibration,
            "games": {"logged": len(games), "graded": len(finals),
                      "pending": sum(r["result"] == "Pending" for r in games),
                      "winner": rate(winner, lambda r: (r["margin"] > 0) == (r["actual_margin"] > 0)),
                      "ats": rate(ats, lambda r: (r["margin"]+r["market_spread"] > 0) == (r["actual_margin"]+r["market_spread"] > 0)),
                      "totals": rate(totals, lambda r: (r["total"]-r["market_total"] > 0) == (r["actual_total"]-r["market_total"] > 0)),
                      "margin_comparison": {"n": len(matched_margin),
                          "model_mae": mean([abs(r["margin"]-r["actual_margin"]) for r in matched_margin]),
                          "market_mae": mean([abs(-r["market_spread"]-r["actual_margin"]) for r in matched_margin])},
                      "total_comparison": {"n": len(matched_total),
                          "model_mae": mean([abs(r["total"]-r["actual_total"]) for r in matched_total]),
                          "market_mae": mean([abs(r["market_total"]-r["actual_total"]) for r in matched_total])},
                      "margin_mae": mean([abs(r["margin"]-r["actual_margin"]) for r in finals if number(r.get("margin")) is not None]),
                      "total_mae": mean([abs(r["total"]-r["actual_total"]) for r in finals if number(r.get("total")) is not None]),
                      "brier": mean([(r["probability"]-(r["actual_margin"] > 0))**2 for r in finals
                                     if number(r.get("probability")) is not None and r["actual_margin"] != 0])},
            "props": {k: {"logged": len(v), "graded": sum(r["result"] == "Graded" for r in v),
                          "mae": mean([abs(r["projection"]-r["actual"]) for r in v if r["result"] == "Graded"]),
                          "bias": mean([r["projection"]-r["actual"] for r in v if r["result"] == "Graded"])} for k,v in sorted(prop_groups.items())},
            "records": sorted(rows, key=lambda r: (r.get("start") or "", r["id"]), reverse=True)}

File: projects/ladderbet/docs/model-accuracy.js
/* Read-only prediction accuracy. Does not read or write any wager ledger. */
(function (root) {
  "use strict";
  const esc = value => String(value ?? "").replace(/[&<>"']/g, c => ({"&":"&amp;","<":"&lt;",">":"&gt;",'"':"&quot;","'":"&#39;"}[c]));
  const num = (v, n=2) => v == null || !Number.isFinite(Number(v)) ? "—" : Number(v).toFixed(n);
  const pct = v => v == null ? "—" : `${num(v*100,1)}%`;
  const caches = new Map();
  function groupTable(title, groups) {
    return `<h3>${esc(title)}</h3><div class="ma-scroll"><table><thead><tr><th>Group</th><th>Logged</th><th>W–L–P</th><th>Pending</th><th>Win %</th><th>Units</th><th>ROI</th><th>Brier</th></tr></thead><tbody>${Object.entries(groups||{}).map(([k,v])=>`<tr><td>${esc(k)}</td><td>${v.logged}</td><td>${v.wins}–${v.losses}–${v.pushes}</td><td>${v.pending}</td><td>${pct(v.win_rate)}</td><td>${num(v.units)}</td><td>${pct(v.roi)}</td><td>${num(v.brier,3)}</td></tr>`).join("") || '<tr><td colspan="8">No predictions recorded yet.</td></tr>'}</tbody></table></div>`;
  }
  function render(host, data) {
    const o=data.overall||{}, g=data.games||{}, rows=data.records||[];
    const props=Object.entries(data.props||{});
    host.classList.add("model-accuracy");
    host.innerHTML=`<style>.model-accuracy{color:inherit;margin:18px 0}.model-accuracy p{line-height:1.6}.model-accuracy h3{margin:22px 0 10px}.ma-grid{display:flex;gap:12px;flex-wrap:wrap}.ma-stat{padding:12px 16px;border:1px solid #8885;border-radius:8px;min-width:130px}.ma-stat b{display:block;font-size:23px}.ma-scroll{overflow:auto;max-height:580px;border:1px solid #8885;border-radius:8px}.model-accuracy table{border-collapse:collapse;width:100%;font-size:13px;min-width:680px}.model-accuracy th,.model-accuracy td{text-align:left;padding:10px;border-bottom:1px solid #8883;white-space:normal}.model-accuracy th{font-weight:700}.ma-controls{display:flex;gap:10px;flex-wrap:wrap;align-items:center;margin:12px 0}.model-accuracy select,.model-accuracy input,.model-accuracy button{font:inherit;color:inherit;background:transparent;border:1px solid #8888;border-radius:5px;padding:7px}.model-accuracy option{color:#222;background:#fff}.model-accuracy small{opacity:.75}.ma-ok{color:#299c69}.ma-loss{color:#d45c55}</style>
      <h2>${esc(data.label||"Model accuracy")}</h2><p>${esc(data.method)}</p>
      <p><small>Updated ${esc(data.generated_at)} · <a href="data/accuracy.json" download>Download full prediction history</a>. Win rates exclude pushes and voids. Brier is probability error; lower is better. Tier versions are kept separate below.</small></p>
      <div class="ma-grid"><div class="ma-stat">Model picks logged<b>${o.logged||0}</b>${o.pending||0} pending</div><div class="ma-stat">Pick record<b>${o.wins||0}–${o.losses||0}–${o.pushes||0}</b>${o.voids||0} void · ${pct(o.win_rate)} wins</div><div class="ma-stat">Flat one-unit result<b>${num(o.units)}</b>${pct(o.roi)} hypothetical ROI</div><div class="ma-stat">Pick Brier<b>${num(o.brier,3)}</b>All selected tiers</div></div>
      ${g.logged ? `<h3>Every game forecast, with or without a bet</h3><div class="ma-grid"><div class="ma-stat">Games<b>${g.graded||0} graded</b>${g.pending||0} pending / ${g.logged} logged</div><div class="ma-stat">Winner accuracy<b>${pct(g.winner?.accuracy)}</b>${g.winner?.correct||0} / ${g.winner?.n||0} decided</div><div class="ma-stat">Spread direction<b>${pct(g.ats?.accuracy)}</b>${g.ats?.n||0} decided</div><div class="ma-stat">Total direction<b>${pct(g.totals?.accuracy)}</b>${g.totals?.n||0} decided</div><div class="ma-stat">Margin / total MAE<b>${num(g.margin_mae,1)} / ${num(g.total_mae,1)}</b>Points of absolute error</div><div class="ma-stat">Margin: model / market<b>${num(g.margin_comparison?.model_mae,1)} / ${num(g.margin_comparison?.market_mae,1)}</b>${g.margin_comparison?.n||0} matched games</div><div class="ma-stat">Total: model / market<b>${num(g.total_comparison?.model_mae,1)} / ${num(g.total_comparison?.market_mae,1)}</b>${g.total_comparison?.n||0} matched games</div></div>`:""}
      ${props.length ? `<h3>Player projections, including unpriced markets</h3><div class="ma-scroll"><table><thead><tr><th>Market</th><th>Logged</th><th>Graded</th><th>Mean absolute error</th><th>Bias (forecast − actual)</th></tr></thead><tbody>${props.map(([k,v])=>`<tr><td>${esc(k)}</td><td>${v.logged}</td><td>${v.graded}</td><td>${num(v.mae)}</td><td>${num(v.bias)}</td></tr>`).join("")}</tbody></table></div>`:""}
      ${!o.settled&&!g.graded&&!props.some(([,v])=>v.graded) ? '<p>Predictions are being saved before games start. Results appear after final scores or final player statistics arrive. Past games without a saved pregame forecast are not counted.</p>':""}
      ${groupTable("Accuracy by original tier",data.by_tier)}${groupTable("Accuracy by market",data.by_market)}
      <details><summary>Probability calibration and tier versions</summary>${groupTable("Tier calculation version",data.by_version)}<div class="ma-scroll"><table><thead><tr><th>Probability bucket</th><th>Sample</th><th>Mean prediction</th><th>Actual win rate</th></tr></thead><tbody>${(data.calibration||[]).map(c=>`<tr><td>${esc(c.bucket)}</td><td>${c.n}</td><td>${pct(c.predicted)}</td><td>${pct(c.actual)}</td></tr>`).join("")||'<tr><td colspan="4">Waiting for decided picks.</td></tr>'}</tbody></table></div></details>
      <h3>Prediction history</h3><p>Game forecasts and preferred market picks are shown by default. The complete audit also includes the opposite priced sides; its combined win rate is not a measure of pick quality.</p>
      <div class="ma-controls"><label>Show <select data-ma-kind><option value="preferred">Forecasts + preferred picks</option value="all">Complete priced-side audit</option value="game">Game forecasts</option><option value="prop">Player forecasts</option><option value="call">Preferred market picks</option></select></label><label>Result <select data-ma-result><option value="all">All</option>${["Pending","Win","Loss","Push","Void","Graded"].map(x=>`<option>${x}</option>`).join("")}</select></label><label>Tier <select data-ma-tier><option value="all">All</option>${[...new Set(rows.map(r=>r.tier).filter(Boolean))].sort().map(x=>`<option>${esc(x)}</option>`).join("")}</select></label><label>Search <input data-ma-search placeholder="Team, player or date" type="search"></label></div>
      <div class="ma-scroll"><table><thead><tr><th>Start / saved</th><th>Forecast or pick</th><th>Original tier</th><th>Prediction</th><th>Actual</th><th>Result</th></tr></thead><tbody data-ma-rows></tbody></table></div><div class="ma-controls"><button data-ma-prev>Previous</button><span data-ma-page></span><button data-ma-next>Next</button></div>`;
    let page=0;
    const q=s=>host.querySelector(s);
    function draw(){
      const kind=q("[data-ma-kind]").value, result=q("[data-ma-result]").value, tier=q("[data-ma-tier]").value, search=q("[data-ma-search]").value.toLowerCase();
      const filtered=rows.filter(r=>(kind==="all"||kind==="preferred"?(kind==="all"||r.kind!=="call"||r.selected):r.kind===kind&&(kind!=="call"||r.selected))&&(result==="all"||r.result===result)&&(tier==="all"||r.tier===tier)&&`${r.matchup} ${r.player||""} ${r.pick||""} ${r.start}`.toLowerCase().includes(search));
      page=Math.min(page,Math.max(0,Math.ceil(filtered.length/50)-1));
      q("[data-ma-rows]").innerHTML=filtered.slice(page*50,page*50+50).map(r=>{
        const prediction=r.kind==="game"?`Home margin ${num(r.margin,1)} · total ${num(r.total,1)}`:r.kind==="prop"?num(r.projection):`${pct(r.probability)} · ${r.price>0?"+":""}${num(r.price,0)}${r.line==null?"":` · line ${num(r.line,1)}`}`;
        const actual=r.actual!=null?num(r.actual):r.final_home!=null?`Away ${num(r.final_away,0)} – home ${num(r.final_home,0)}`:"—";
        return `<tr><td>${esc(r.start)}<br><small>Saved ${esc(r.captured_at)}</small></td><td>${esc(r.matchup)}<br>${esc(r.pick||[r.player,r.market].filter(Boolean).join(" · ")||"Game forecast")}</td><td>${esc(r.tier||"Forecast")}<br><small>${esc(r.version)}</small></td><td>${esc(prediction)}</td><td>${esc(actual)}</td><td class="${r.result==="Win"?"ma-ok":r.result==="Loss"?"ma-loss":""}">${esc(r.result)}${r.units==null?"":`<br>${num(r.units)} units`}</td></tr>`;
      }).join("")||'<tr><td colspan="6">No matching predictions.</td></tr>';
      q("[data-ma-page]").textContent=`${filtered.length} records · page ${page+1} of ${Math.max(1,Math.ceil(filtered.length/50))}`;
      q("[data-ma-prev]").disabled=page===0; q("[data-ma-next]").disabled=(page+1)*50>=filtered.length;
    }
    host.querySelectorAll("select,input").forEach(e=>e.addEventListener("input",()=>{page=0;draw();}));
    q("[data-ma-prev]").onclick=()=>{page--;draw();};q("[data-ma-next]").onclick=()=>{page++;draw();};draw();
  }
  let routed = false;
  function scan(){
    if (!routed && new URLSearchParams(location.search).get("tab") === "accuracy") {
      const button = [...document.querySelectorAll("button")].find(b=>/^Accuracy/.test(b.textContent.trim()));
      if (button) { routed = true; button.click(); }
    }
    document.querySelectorAll("[data-model-accuracy]").forEach(host=>{
      if(host.dataset.accuracyMounted)return;
      host.dataset.accuracyMounted="1";host.textContent="Loading prediction history…";
      const url=host.dataset.modelAccuracy||"data/accuracy.json";
      if(!caches.has(url))caches.set(url,fetch(url,{cache:"no-cache"}).then(r=>{if(!r.ok)throw new Error("unavailable");return r.json();}));
      caches.get(url).then(d=>render(host,d)).catch(()=>{host.textContent="Prediction history is unavailable. Check the latest refresh status and reload.";});
    });
  }
  if(typeof document!=="undefined"){
    if(document.readyState==="loading")document.addEventListener("DOMContentLoaded",scan);else scan();
    new MutationObserver(scan).observe(document.documentElement,{childList:true,subtree:true});
  }
  if(typeof module!=="undefined")module.exports={esc,num,pct};
})(typeof globalThis!=="undefined"?globalThis:this);
